llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -1687,6 +1687,44 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
   return nullptr;
 }
 
+static Value *foldSelectInstWithICmpOr(SelectInst &SI, ICmpInst *ICI,
+                                       InstCombiner::BuilderTy &Builder) {
+
+  // a > -1 ? 1 : (a | value) --> smin(1, a | value)
+  // a >= 0 ? 1 : (a | value) --> smin(1, a | value)
+  const APInt *Cmp;
+  Value *A = ICI->getOperand(0);
+  Value *B = ICI->getOperand(1);
+  Type *Ty = SI.getType();
+  Value *TVal = SI.getTrueValue();
+  Value *FVal = SI.getFalseValue();
+  Constant *One = ConstantInt::get(Ty, 1);
+  Constant *NegOne = ConstantInt::get(Ty, -1);
+  Constant *Zero = Constant::getNullValue(Ty);
+  CmpInst::Predicate Pred = ICI->getPredicate();
+
+  if (!match(B, m_APIntAllowUndef(Cmp)))
+    return nullptr;
+
+  // Swap TVal, FVal for Inverse
+  if (Pred == ICmpInst::ICMP_SLT || Pred == ICmpInst::ICMP_SLE)
+    std::swap(TVal, FVal);
+
+  if (!(match(TVal, m_One()) || match(TVal, m_Zero()) ||
+        match(TVal, m_AllOnes())) ||
+      !match(FVal, m_Or(m_Value(A), m_StrictlyPositive())))
+    return nullptr;
+
+  if (((Pred == ICmpInst::ICMP_SGT && *Cmp == -1) ||
+       (Pred == ICmpInst::ICMP_SGE && *Cmp == 0)))
+    if (match(TVal, m_One()))
+      return Builder.CreateBinaryIntrinsic(Intrinsic::smin, One, FVal);
+  if (match(TVal, m_Zero()))
+    return Builder.CreateBinaryIntrinsic(Intrinsic::smin, Zero, FVal);
+  if (match(TVal, m_AllOnes()))
+    return Builder.CreateBinaryIntrinsic(Intrinsic::smin, NegOne, FVal);
```

Comment:
This could be simplified to:
```
if(match(TVal, m_One()) || match(TVal, m_AllOnes()) || match(TVal, m_Zero())) {
  return Builder.Create...(smin, TVal, FVal);
}
```

Reply:
Also your comment / proofs only seem to cover the `TVal == 1` case, can you please update them to cover the other two cases.